pyro/infer/predictive.py

```diff
@@ -450,3 +450,181 @@ def forward(self, *args, **kwargs):
             guide_log_prob=guide_log_prob,
             model_log_prob=model_log_prob,
         )
+
+
+class MHResampler(torch.nn.Module):
+    r"""
+    Resampler for weighed samples that generates equally weighed samples from the distribution
+    specified by the weighed samples ``sampler``.
+
+    The resampling is based on the Metropolis-Hastings algorithm.
+    Given an initial sample :math:`x` subsequent samples are generated by:
+
+    -   Sampling from the ``guide`` a new sample candidate :math:`x'` with probability :math:`g(x')`.
+    -   Calculate an acceptance probability
+        :math:`A(x', x) = \min\left(1, \frac{P(x')}{P(x)} \frac{g(x)}{g(x')}\right)`
+        with :math:`P` being the ``model``.
+    -   With probability :math:`A(x', x)` accept the new sample candidate :math:`x'`
+        as the next sample, otherwise set the current sample :math:`x` as the next sample.
+
+    The above is the Metropolis-Hastings algorithm with the new sample candidate
+    proposal distribution being equal to the ``guide`` and independent of the
+    current sample such that :math:`g(x')=g(x' \mid x)`.
+
+    :param callable sampler: When called returns :class:`WeighedPredictiveResults`.
+    :param slice source_samples_slice: Select source samples for storage (default is `slice(0)`, i.e. none).
+    :param slice stored_samples_slice: Select output samples for storage (default is `slice(0)`, i.e. none).
+
+    The typical use case of :class:`MHResampler` would be to convert weighed samples
+    generated by :class:`WeighedPredictive` into equally weighed samples from the target distribution.
```

Comment:
Could you add a phrase or sentence regarding the intended iterative use of this function? Is it intended to be called repeatedly? What happens if you call it only once? (see my question below)

Reply:
I've added the below to the docs:

Each time an instance of `MHResampler` is called it returns a new set of samples, with the samples generated by the first call being distributed according to the ``guide``, and with each subsequent call the distribution of the samples becomes closer to that of the posterior predictive disdtribution. It might take some experimentation in order to find out in each case how many times one would need to call an instance of `MHResampler` in order to be close enough to the posterior predictive distribution.